### PR TITLE
Improve code coverage

### PR DESCRIPTION
**Pull request by chenzhitong**

### CODE REVIEW

File: tests/Neo.Plugins.Storage.Tests/libleveldb.dll
Comment (cschuchardt88) [inline tests/Neo.Plugins.Storage.Tests/libleveldb.dll:1]: We don't allow binary files in the repo. #3351 fixes issue with this file for nuget package ***only***.

**Review by cschuchardt88 — CHANGES_REQUESTED** [review]

**Review by cschuchardt88 — APPROVED** [review]

**Comment by chenzhitong** [comment]
Thanks @cschuchardt88 , I'll keep adding UT

File: src/Neo/SmartContract/Manifest/ContractAbi.cs [hunk 1]
```diff
@@ -62,10 +62,10 @@ public static ContractAbi FromJson(JObject json)
         {
             ContractAbi abi = new()
             {
-                Methods = ((JArray)json["methods"]).Select(u => ContractMethodDescriptor.FromJson((JObject)u)).ToArray(),
-                Events = ((JArray)json["events"]).Select(u => ContractEventDescriptor.FromJson((JObject)u)).ToArray()
+                Methods = ((JArray)json?["methods"])?.Select(u => ContractMethodDescriptor.FromJson((JObject)u)).ToArray(),
```
[the remark this answers is not in the document]
Comment (chenzhitong) [reply src/Neo/SmartContract/Manifest/ContractAbi.cs:65]: Followed by a line of code 
`if (abi.Methods == null || abi.Methods.Length == 0) throw new FormatException();`
 I add `?` to avoid null reference exceptions. When adding unit tests, I found that this is prone to null reference exceptions

File: src/Neo/Cryptography/Base58.cs [hunk 2]
```diff
@@ -34,7 +34,7 @@ public static class Base58
         /// <returns>A byte array that is equivalent to <paramref name="input"/>.</returns>
         public static byte[] Base58CheckDecode(this string input)
         {
-            if (input is null) throw new ArgumentNullException(nameof(input));
+            if (string.IsNullOrEmpty(input)) throw new ArgumentException(nameof(input));
```
Comment (shargon) [inline src/Neo/Cryptography/Base58.cs:37]: Why change the exception?

File: src/Neo/SmartContract/Manifest/ContractPermissionDescriptor.cs [hunk 3]
```diff
@@ -114,7 +114,8 @@ public bool Equals(ContractPermissionDescriptor other)
             if (this == other) return true;
             if (IsWildcard == other.IsWildcard) return true;
             if (IsHash) return Hash.Equals(other.Hash);
-            else return Group.Equals(other.Group);
+            if (IsGroup) return Group.Equals(other.Group);
```
Comment (shargon) [inline src/Neo/SmartContract/Manifest/ContractPermissionDescriptor.cs:117]: The logic changed, not only was added UT

File: tests/Neo.UnitTests/Neo.UnitTests.csproj [hunk 4]
```diff
@@ -12,6 +12,9 @@
   </ItemGroup>
 
   <ItemGroup>
+    <None Update="SmartContract\Manifest\TestFile\**">
```
Comment (shargon) [inline tests/Neo.UnitTests/Neo.UnitTests.csproj:15]:
```suggestion
    <None Update="SmartContract\Manifest\Artifacts\**">
```

use artifacts folder?

File: src/Neo/Cryptography/Base58.cs (hunk 2 above)
Comment (chenzhitong) [reply src/Neo/Cryptography/Base58.cs:37]: Because the `string input` parameter is always non-null.

File: src/Neo/SmartContract/Manifest/ContractPermissionDescriptor.cs (hunk 3 above)
Comment (chenzhitong) [reply src/Neo/SmartContract/Manifest/ContractPermissionDescriptor.cs:117]: Because in the original code, the `Group` may be null, and there will be an exception

File: tests/Neo.UnitTests/Neo.UnitTests.csproj (hunk 4 above)
Comment (chenzhitong) [reply tests/Neo.UnitTests/Neo.UnitTests.csproj:15]: Artifacts is in .gitignore, I can't add files to it.

File: src/Neo/Cryptography/Base58.cs (hunk 2 above)
Comment (cschuchardt88) [reply src/Neo/Cryptography/Base58.cs:37]: Its not `non-null`. Than you should be doing `string.IsNullOrWhitespace`

**Review by cschuchardt88 — APPROVED** [review]

**Comment by cschuchardt88** [comment]
If you could please update the description of the `1st comment` of what this PR covers. So when we merge, we will know.

File: src/Neo/SmartContract/Manifest/ContractManifest.cs [hunk 5]
```diff
@@ -110,22 +110,27 @@ public StackItem ToStackItem(ReferenceCounter referenceCounter)
...
             {
-                Name = json["name"].GetString(),
-                Groups = ((JArray)json["groups"]).Select(u => ContractGroup.FromJson((JObject)u)).ToArray(),
-                SupportedStandards = ((JArray)json["supportedstandards"]).Select(u => u.GetString()).ToArray(),
-                Abi = ContractAbi.FromJson((JObject)json["abi"]),
-                Permissions = ((JArray)json["permissions"]).Select(u => ContractPermission.FromJson((JObject)u)).ToArray(),
-                Trusts = WildcardContainer<ContractPermissionDescriptor>.FromJson(json["trusts"], u => ContractPermissionDescriptor.FromJson((JString)u)),
-                Extra = (JObject)json["extra"]
-            };
+                manifest.Name = json["name"]?.GetString();
```
Comment (shargon) [inline src/Neo/SmartContract/Manifest/ContractManifest.cs:116]: previously fail when null, now set name as null

**Review by shargon — CHANGES_REQUESTED** [review]

**Comment by Jim8y** [comment]
I need some time to ensure that your update will not change existing behavior. for instance, if the original code should throw exception, it should not return null in your code.

File: src/Neo/SmartContract/Manifest/ContractManifest.cs (hunk 5 above)
Comment (chenzhitong) [reply src/Neo/SmartContract/Manifest/ContractManifest.cs:116]: previously crash when null, now it throws a FormatException.

Comment (chenzhitong) [reply src/Neo/SmartContract/Manifest/ContractManifest.cs:116]: previously crash when null, now it throws a FormatException. Demo:
image

Comment (cschuchardt88) [reply src/Neo/SmartContract/Manifest/ContractManifest.cs:116]: `name` is required. 

https://github.com/neo-project/proposals/blob/master/nep-15.mediawiki#user-content-Manifest

File: src/Neo/SmartContract/Manifest/ContractManifest.cs [hunk 6]
```diff
@@ -110,22 +110,27 @@ public StackItem ToStackItem(ReferenceCounter referenceCounter)
...
+                manifest.Groups = ((JArray)json["groups"])?.Select(u => ContractGroup.FromJson((JObject)u)).ToArray();
```
Comment (cschuchardt88) [inline src/Neo/SmartContract/Manifest/ContractManifest.cs:117]:
```suggestion
                manifest.Groups = ((JArray)json["groups"])?.Select(u => ContractGroup.FromJson((JObject)u)).ToArray() ?? [];
```

File: src/Neo/SmartContract/Manifest/ContractManifest.cs [hunk 7]
```diff
@@ -110,22 +110,27 @@ public StackItem ToStackItem(ReferenceCounter referenceCounter)
...
+                manifest.SupportedStandards = ((JArray)json["supportedstandards"])?.Select(u => u.GetString()).ToArray();
```
Comment (cschuchardt88) [inline src/Neo/SmartContract/Manifest/ContractManifest.cs:118]:
```suggestion
                manifest.SupportedStandards = ((JArray)json["supportedstandards"])?.Select(u => u.GetString()).ToArray() ?? [];
```

File: src/Neo/SmartContract/Manifest/ContractManifest.cs [hunk 8]
```diff
@@ -110,22 +110,27 @@ public StackItem ToStackItem(ReferenceCounter referenceCounter)
...
+                manifest.Abi = ContractAbi.FromJson((JObject)json["abi"]);
+                manifest.Permissions = ((JArray)json["permissions"])?.Select(u => ContractPermission.FromJson((JObject)u)).ToArray();
```
Comment (cschuchardt88) [inline src/Neo/SmartContract/Manifest/ContractManifest.cs:120]:
```suggestion
                manifest.Permissions = ((JArray)json["permissions"])?.Select(u => ContractPermission.FromJson((JObject)u)).ToArray() ?? [];
```

**Review by cschuchardt88 — CHANGES_REQUESTED** [review]

File: src/Neo/SmartContract/Manifest/ContractManifest.cs (hunk 5 above)
Comment (shargon) [reply src/Neo/SmartContract/Manifest/ContractManifest.cs:116]: but this maybe can produce a fork

Comment (Jim8y) [reply src/Neo/SmartContract/Manifest/ContractManifest.cs:116]: > previously crash when null, now it throws a FormatException. Demo:
> image
> 

a big no here. null and exception are totally different behavior.

Comment (chenzhitong) [reply src/Neo/SmartContract/Manifest/ContractManifest.cs:116]: I just fixed the bug where the wrong json was causing a crash. Previously the node would simply crash without throwing an exception.

Comment (chenzhitong) [reply src/Neo/SmartContract/Manifest/ContractManifest.cs:116]: > `name` is required.
> 
> https://github.com/neo-project/proposals/blob/master/nep-15.mediawiki#user-content-Manifest

Yes, manifest.Name is not allowed to be null, there's another line of code after it (it was there before, not a new one I added)
So it won't cause a fork @shargon @cschuchardt88 
```
if (string.IsNullOrEmpty(manifest.Name))
    throw new FormatException();
```

Comment (shargon) [reply src/Neo/SmartContract/Manifest/ContractManifest.cs:116]: then use `!` instead of `?` to avoid the behavior change

File: src/Neo/SmartContract/Manifest/ContractPermissionDescriptor.cs (hunk 3 above)
Comment (shargon) [reply src/Neo/SmartContract/Manifest/ContractPermissionDescriptor.cs:117]: This behaviour change should be done in a different PR

**Comment by lingyido** [comment]
`Base58CheckDecode`'s check may cause hard fork, we'd better have an empty string tested here on both older version and newer version

**Comment by chenzhitong** [comment]
> `Base58CheckDecode`'s check may cause hard fork, we'd better have an empty string tested here on both older version and newer version

In both older and newer versions, `Base58CheckDecode` throws an exception when an empty string is entered.

image

In the old version, `if(input is null)` was always `false`, because the parameter type is `string`, not `string?`

**Comment by lingyido** [comment]
> In the old version, `if(input is null)` was always `false`, because the parameter type is `string`, not `string?`

I thought that the empty string `""` may cause differ results in the two versions. In previous version it will just return empty byte array `new byte[]{}` while it will throw exception in this new version?

There should comes a test for this I guess.

**Comment by cschuchardt88** [comment]
Yes it will be detected by the `ApplicationEngine`.

**Comment by chenzhitong** [comment]
Just in case, the exception type will not be modified here

File: src/Neo/Cryptography/Base58.cs [hunk 9]
```diff
@@ -34,7 +34,7 @@ public static class Base58
         /// <returns>A byte array that is equivalent to <paramref name="input"/>.</returns>
         public static byte[] Base58CheckDecode(this string input)
         {
-            if (string.IsNullOrEmpty(input)) throw new ArgumentException(nameof(input));
+            if (string.IsNullOrEmpty(input)) throw new FormatException(nameof(input));
```
Comment (cschuchardt88) [inline src/Neo/Cryptography/Base58.cs:37]: `FormatException` exception should be `inner` exception of `ArgumentException`.

File: src/Neo/SmartContract/Manifest/ContractManifest.cs (hunk 5 above)
Comment (chenzhitong) [reply src/Neo/SmartContract/Manifest/ContractManifest.cs]: fixed

File: src/Neo/SmartContract/Manifest/ContractPermissionDescriptor.cs (hunk 3 above)
Comment (chenzhitong) [reply src/Neo/SmartContract/Manifest/ContractPermissionDescriptor.cs]: fixed

Comment (chenzhitong) [reply src/Neo/SmartContract/Manifest/ContractPermissionDescriptor.cs]: https://github.com/neo-project/neo/pull/3396

**Review by shargon — APPROVED** [review]

**Review by Jim8y — APPROVED** [review]

File: src/Neo/SmartContract/Manifest/ContractAbi.cs [hunk 10]
```diff
@@ -62,8 +62,8 @@ public static ContractAbi FromJson(JObject json)
         {
             ContractAbi abi = new()
             {
-                Methods = ((JArray)json["methods"]).Select(u => ContractMethodDescriptor.FromJson((JObject)u)).ToArray(),
-                Events = ((JArray)json["events"]).Select(u => ContractEventDescriptor.FromJson((JObject)u)).ToArray()
+                Methods = ((JArray)json!["methods"])!.Select(u => ContractMethodDescriptor.FromJson((JObject)u)).ToArray(),
+                Events = ((JArray)json!["events"])!.Select(u => ContractEventDescriptor.FromJson((JObject)u)).ToArray()
```
Comment (cschuchardt88) [inline src/Neo/SmartContract/Manifest/ContractAbi.cs:66]:
```suggestion
                Methods = ((JArray)json?["methods"])?.Select(u => ContractMethodDescriptor.FromJson((JObject)u)).ToArray() ?? [],
                Events = ((JArray)json?["events"])?.Select(u => ContractEventDescriptor.FromJson((JObject)u)).ToArray() ?? []
```

File: src/Neo/SmartContract/Manifest/ContractManifest.cs [hunk 11]
```diff
@@ -112,20 +112,21 @@ public static ContractManifest FromJson(JObject json)
         {
             ContractManifest manifest = new()
             {
-                Name = json["name"].GetString(),
-                Groups = ((JArray)json["groups"]).Select(u => ContractGroup.FromJson((JObject)u)).ToArray(),
-                SupportedStandards = ((JArray)json["supportedstandards"]).Select(u => u.GetString()).ToArray(),
+                Name = json["name"]!.GetString(),
+                Groups = ((JArray)json["groups"])!.Select(u => ContractGroup.FromJson((JObject)u)).ToArray(),
+                SupportedStandards = ((JArray)json["supportedstandards"])!.Select(u => u.GetString()).ToArray(),
```
Comment (cschuchardt88) [inline src/Neo/SmartContract/Manifest/ContractManifest.cs:117]:
```suggestion
                Groups = ((JArray)json?["groups"])?.Select(u => ContractGroup.FromJson((JObject)u)).ToArray() ?? [],
                SupportedStandards = ((JArray)json?["supportedstandards"])?.Select(u => u.GetString()).ToArray() ?? [],
```

File: src/Neo/SmartContract/Manifest/ContractManifest.cs [hunk 12]
```diff
@@ -112,20 +112,21 @@ public static ContractManifest FromJson(JObject json)
...
                 Abi = ContractAbi.FromJson((JObject)json["abi"]),
-                Permissions = ((JArray)json["permissions"]).Select(u => ContractPermission.FromJson((JObject)u)).ToArray(),
+                Permissions = ((JArray)json["permissions"])!.Select(u => ContractPermission.FromJson((JObject)u)).ToArray(),
```
Comment (cschuchardt88) [inline src/Neo/SmartContract/Manifest/ContractManifest.cs:119]:
```suggestion
                Permissions = ((JArray)json?["permissions"])?.Select(u => ContractPermission.FromJson((JObject)u)).ToArray() ?? [],
```

**Comment by cschuchardt88** [comment]
Not ready until changes are made.

File: src/Neo/SmartContract/Manifest/ContractAbi.cs (hunk 10 above)
Comment (cschuchardt88) [inline src/Neo/SmartContract/Manifest/ContractAbi.cs:66]:
```suggestion
                Methods = ((JArray)json!["methods"])?.Select(u => ContractMethodDescriptor.FromJson((JObject)u)).ToArray() ?? [],
                Events = ((JArray)json!["events"])?.Select(u => ContractEventDescriptor.FromJson((JObject)u)).ToArray() ?? []
```

You dont use `TryCatch` here. So I fixed it.

**Review by cschuchardt88 — APPROVED** [review]

**Review by shargon — APPROVED** [review]

**Comment by shargon** [comment]
@Jim8y do you want to check the last changes?
